lib/forwardanalyzer.cpp

```diff
@@ -585,6 +585,31 @@ struct ForwardTraversal {
             } else if (Token::simpleMatch(tok, "switch (")) {
                 if (updateRecursive(tok->next()->astOperand2()) == Progress::Break)
                     return Break();
+                if (!analyzer->isKnown())
```

Comment:
This should use `analyzer->evaluate(tok->next()->astOperand2())`, and check if the vector is empty.

Reply:
Also, we should still traverse the switch statement even if the value is not known.

`analyzer->evaluate` will evaluate the switch value, and then we can select the case statement of the switch value. When we can't evaluate the switch, we should stop if the value `isConditional`. When its not conditional then we need to analyze all of the cases(with `analyzeRange`). If some of the cases are modified then we lower to possible, then we fork on the cases that modify the value(we should probably extend `forkRange` to return the `ForwardTraversal`) and continue analysis on the cases that do not modify the value.

Of course, we dont need to do this all at once. We can start with propagating values for when the switch can be evaluated or when the values are never modified in the switch body.

Reply:
I tested with 1588e6364cf314cb86967ffca69d9671ba166fb3 and it does not work fully. So I probably misunderstood what I should do..

Reply:
Normally I would do this(after your `if` statement):

```cpp
                Token *bodyStart = tok->linkAt(1)->next();
                const Token *bodyEnd = bodyStart->link();
                Analyzer::Action a = analyzeScope(bodyEnd);
                if (a.isModified() && !analyzer->lowerToPossible())
                    return Break();
                if (a.isInconclusive() && !analyzer->lowerToInconclusive())
                    return Break();
                if (updateRange(bodyStart, bodyEnd) == Progress::Break)
                    return Break();
                tok = endBlock;
```

But this wont work fully for two reasons. One it will stop when we hit `break` or `return`. Second, it will lowerToPossible when it sees a case label. Instead, I think it would be better to gather all the cases(we can expand it later to a map from case value to range):

```cpp
                std::vector<std::pair<Token*, Token*>> cases = gatherCases(bodyEnd);
                int invalidCases = 0;
                Analyzer::Action action;
                for(auto&& pcase:cases) {
                    bool breakCase = !Token::Match(pcase.second->tokAt(-2), "break ;");
                    Analyzer::Action a = analyzeRange(pcase.first, pcase.second);
                    // Need to update forkRange to return ForwardTraversal
                    ForwardTraversal ft = forkRange(pcase.first, pcase.second);
                    if(ft.terminate == Terminate::Escape && !breakCase) {
                        invalidCases++;
                    }
                    // Only update the action when its non-terminating
                    else if(a.isModified() || a.isInconclusive()) {
                        action |= a;
                        invalidCases++;
                    }
                }
                if(invalidCases == cases.size())
                    return Break();
                if (action.isModified() && !analyzer->lowerToPossible())
                    return Break();
                if (action.isInconclusive() && !analyzer->lowerToInconclusive())
                    return Break();
                tok = endBlock;
```

Although, I haven't tested any of this code, so this is just a rough idea how I would tackle this problem.